aardvark/persistence/sqlalchemy/__init__.py

```diff
@@ -0,0 +1,304 @@
+            if isinstance(last_authenticated, int):
+                dt_last_authenticated = datetime.datetime.fromtimestamp(
+                    last_authenticated / 1e3
+                )
+            elif isinstance(last_authenticated, str):
+                dt_last_authenticated = datetime.datetime.strptime(
+                    last_authenticated, "%Y-%m-%d %H:%M:%S.%f"
+                )
+            else:
+                dt_last_authenticated = last_authenticated
+
+            dt_starting = datetime.datetime.utcnow() - datetime.timedelta(days=90)
+            usage[namespace]["USED_LAST_90_DAYS"] = dt_last_authenticated > dt_starting
+
+        return usage
+
+    def store_role_data(self, access_advisor_data: Dict[str, Any]):
+        with self.session_scope() as session:
+            if not access_advisor_data:
+                log.warning(
+                    "Cannot persist Access Advisor Data as no data was collected."
+                )
+                return
+
+            arn_cache = {}
+            for arn, data in access_advisor_data.items():
+                if arn in arn_cache:
+                    item = arn_cache[arn]
+                else:
+                    item = self.get_or_create_iam_object(arn)
+                    arn_cache[arn] = item
+                for service in data:
+                    self.create_or_update_advisor_data(
+                        item.id,
+                        service["LastAuthenticated"],
+                        service["ServiceName"],
+                        service["ServiceNamespace"],
+                        service.get("LastAuthenticatedEntity"),
+                        service["TotalAuthenticatedEntities"],
+                        session=session,
+                    )
+
+    def get_role_data(
+        self,
+        page: int = 0,
+        count: int = 0,
+        combine: bool = False,
+        phrase: str = "",
+        arns: Optional[List[str]] = None,
+        regex: str = "",
+        session=None,
+    ) -> Dict[str, Any]:
+        offset = (page - 1) * count if page else 0
+        limit = count
+        session = session or self._create_session()
+        # default unfiltered query
+        query = session.query(AWSIAMObject)
+
+        try:
+            if phrase:
+                query = query.filter(AWSIAMObject.arn.ilike("%" + phrase + "%"))
+
+            if arns:
+                query = query.filter(
+                    sa_func.lower(AWSIAMObject.arn).in_([arn.lower() for arn in arns])
+                )
+
+            if regex:
+                query = query.filter(AWSIAMObject.arn.regexp(regex))
+
+            total = query.count()
+
+            if offset:
+                query = query.offset(offset)
+
+            if limit:
+                query = query.limit(limit)
+
+            items = query.all()
+        except Exception as e:
+            raise DatabaseException("Could not retrieve roles from database: %s", e)
+
+        if not items:
```

Comment:
Can you explain more about why we have this failure case?
If the query they ask for returns no results, I feel like we might want to tell them that

Reply:
I can't. This is the existing logic and I was afraid to remove this part before we had more thorough testing in place. It seems like maybe it's a compensation for a lack of validation of the query parameters.